iroh-net/src/endpoint.rs

```diff
@@ -0,0 +1,542 @@
+    pub fn on_endpoints(
+        mut self,
+        on_endpoints: Box<dyn Fn(&[cfg::Endpoint]) + Send + Sync + 'static>,
+    ) -> Self {
+        self.callbacks.on_endpoints = Some(on_endpoints);
+        self
+    }
+
+    /// Optionally provides a func to be called when a connection is made to a DERP server.
+    pub fn on_derp_active(mut self, on_derp_active: Box<dyn Fn() + Send + Sync + 'static>) -> Self {
+        self.callbacks.on_derp_active = Some(on_derp_active);
+        self
+    }
+
+    /// A callback that provides a `cfg::NetInfo` when discovered network conditions change.
+    pub fn on_net_info(
+        mut self,
+        on_net_info: Box<dyn Fn(cfg::NetInfo) + Send + Sync + 'static>,
+    ) -> Self {
+        self.callbacks.on_net_info = Some(on_net_info);
+        self
+    }
+
+    /// Bind the magic endpoint on the specified socket address.
+    pub async fn bind(self, bind_addr: SocketAddr) -> anyhow::Result<MagicEndpoint> {
+        let keypair = self.keypair.unwrap_or_else(Keypair::generate);
+        let mut server_config = make_server_config(
+            &keypair,
+            self.alpn_protocols,
+            self.transport_config,
+            self.keylog,
+        )?;
+        if let Some(c) = self.concurrent_connections {
+            server_config.concurrent_connections(c);
+        }
+        MagicEndpoint::bind(
+            keypair,
+            bind_addr,
+            Some(server_config),
+            self.derp_map,
+            Some(self.callbacks),
+            self.keylog,
+        )
+        .await
+    }
+}
+
+fn make_server_config(
+    keypair: &Keypair,
+    alpn_protocols: Vec<Vec<u8>>,
+    transport_config: Option<quinn::TransportConfig>,
+    keylog: bool,
+) -> anyhow::Result<quinn::ServerConfig> {
+    let tls_server_config = tls::make_server_config(keypair, alpn_protocols, keylog)?;
+    let mut server_config = quinn::ServerConfig::with_crypto(Arc::new(tls_server_config));
+    server_config.transport_config(Arc::new(transport_config.unwrap_or_default()));
+    Ok(server_config)
+}
+
+#[derive(Clone, Debug)]
+pub struct MagicEndpoint {
+    keypair: Arc<Keypair>,
+    conn: Conn,
+    endpoint: quinn::Endpoint,
+    netmap: Arc<Mutex<netmap::NetworkMap>>,
+    keylog: bool,
+}
+
+impl MagicEndpoint {
+    /// Build a MagicEndpoint
+    pub fn builder() -> MagicEndpointBuilder {
+        MagicEndpointBuilder::default()
+    }
+
+    /// Create a quinn endpoint backed by a magicsock.
+    ///
+    /// The *bind_addr* is the address that should be bound locally.  Even though this is an
+    /// outgoing connection a socket must be bound and this is explicit.  The main choice to
+    /// make here is the address family: IPv4 or IPv6.  Otherwise you normally bind to the
+    /// `UNSPECIFIED` address on port `0` thus allowing the kernel to do the right thing.
+    ///
+    /// If *peer_id* is present it will verify during the TLS connection setup that the remote
```

Comment:
not seeing `peer_id` as an argument

Reply:
Ah right, those were copied over from old docs. As the function is pub(crate) I'd move the docs to the builder.